src/geometry/GeometryEvaluator.cc

```diff
@@ -535,12 +536,162 @@ Response GeometryEvaluator::visit(State& state, const RootNode& node)
+		faceNormal.push_back(norm);
+		for(int j=0;j<pol.size(); j++) {
+			Vector3d  pt=pol[j];
+			if(!pointInds.count(pt))
+			{
+				intList newList;
+				newList.push_back(i);
+				pointInds[pt]=newList;
+			}
+			else {
+				pointInds[pt].push_back(i);
+			}
+		}
+	}
+
+	for( const auto& [pt, indexes] : pointInds ) {
+		Vector3d newpt;
+		int valid;
+		do
+		{
+			valid=0;
+			double xmax=-1, ymax=-1, zmax=-1;
+			int xind=-1, yind=-1, zind=-1;
+			
+			// find closest  normal to xdir
+			for(int i=0;i<indexes.size();i++) {
+				if(fabs(faceNormal[indexes[i]][0]) > xmax) {
+					xmax=fabs(faceNormal[indexes[i]][0]);
+					xind=indexes[i];
+				}
+			}
+			if(xind == -1) break;
+
+			// find closest  normal to ydir
+			for(int i=0;i<indexes.size();i++) {
+				if(fabs(faceNormal[indexes[i]][1]) > ymax && faceNormal[indexes[i]].dot(faceNormal[xind]) < 0.999) {
+					ymax=fabs(faceNormal[indexes[i]][1]);
+					yind=indexes[i];
+				}
+			}
+			if(yind == -1) break;
+
+			// find closest  normal to zdir
+			for(int i=0;i<indexes.size();i++) {
+				if(fabs(faceNormal[indexes[i]][2]) > zmax && faceNormal[indexes[i]].dot(faceNormal[xind]) < 0.999 && faceNormal[indexes[i]].dot(faceNormal[yind]) < 0.999 ) {
+					zmax=fabs(faceNormal[indexes[i]][2]);
+					zind=indexes[i];
+				}
+			}
+			if(zind == -1) break;
+			
+			// now calculate the new pt
+			if(cut_face_face_face(
+						pt  +faceNormal[xind]*off  , faceNormal[xind], 
+						pt  +faceNormal[yind]*off  , faceNormal[yind], 
+						pt  +faceNormal[zind]*off  , faceNormal[zind], 
+						newpt)) break;
+			valid=1;
+		} while(0);
+		if(!valid)
+		{
+			Vector3d dir={0,0,0};
+			for(int j=0;j<indexes.size();j++)
+			{
+				if(faceNormal[indexes[j]].norm() > 0.999) dir += faceNormal[indexes[j]];
+			}
+			dir.normalize();
+			newpt  = pt + off* dir;
+		}
+		pointMap[pt]=newpt;
+	}
+
+	// Map all points and assemble
+	PolySet *offset_result = new PolySet(3, /* convex */ true);
+
+	for(int i=0;i<ps->polygons.size();i++) {
+		Polygon offset_polygon;
+		Polygon pol = ps->polygons[i];
+		for(int j=0;j<pol.size(); j++) {
+			Vector3d pt =  pointMap[pol[j]];
+			offset_polygon.push_back(pt);
+		}
+  		offset_result->polygons.push_back(offset_polygon);
+	}
+	return offset_result;
+}
+
 Response GeometryEvaluator::visit(State& state, const OffsetNode& node)
 {
   if (state.isPrefix() && isSmartCached(node)) return Response::PruneTraversal;
   if (state.isPostfix()) {
     shared_ptr<const Geometry> geom;
     if (!isSmartCached(node)) {
+//   unsigned int dim = 0;
+//   int index = node.index();
+//   printf("visited is %p\n",this->visitedchildren);
+//   if(this != NULL) {
+//   for (const auto& item : this->visitedchildren[index]) {
+//      if (!item) {
+//		int   f=item.first();
+//	      printf("t %d %d\n",item.first.getDimension(), item.second.getDimension());
+//      }
+//      }
+//    }
       const Geometry *geometry = applyToChildren2D(node, OpenSCADOperator::UNION);
```

Comment:
Don't call `applyToChildren2D` without checking dimension first.  See `GeometryEvaluator::applyToChildren(const AbstractNode& node, OpenSCADOperator op)` for operations which handle both 2D and 3D.  Probably would be easiest to call that, and add `OFFSET` to `enum OpenSCADOperator` in `enums.h`